CoreCLR interpreter basic compilation and execution

BrzVlad opened the pull request:
In this stage, the implementation does not account for compilation races between multiple threads. Some of the code/functionality will be used in further commits. Additionally there are placeholders for memory allocation, that just use malloc. In the future, most of the memory used during compilation should be cheaply allocated from a compilation memory pool. Some other compilation related memory might need to be freed early to prevent memory waste for huge methods. It would actually make sense for this memory to be allocated with malloc. Other data will need to have its lifetime tied to the MethodDesc lifetime, so if a MethodDesc is ever released, a callback into the interpreter would be necessary to free up this data. These problems are not really relevant in the early stage of the prototype and they can be cleanly addressed later on in the implementation.

The PR is split up in a few commits that can be reviewed independently for convenience.

BrzVlad committed:
Add interpreter test

The test contains one application that contains a method named `RunInterpreterTests` serving as the interpreter entry point. This application is built only, as it is not the test itself. This application is executed with corerun by the actual test, which waits for the corerun exit code to determine success.

BrzVlad committed:
Interpreter wire up

This commit adds a new library for the interpreter. This library is meant to be used as an alt jit compiler, exporting the `getJit` method that returns an interface for method compilation. This library also exports the `getInterpreter` method that returns a separate interface to be used for obtaining any other data from the interpreter library. Method compilation has the following responsability:
        - create an InterpMethod* for the MethodDesc that it flags as compiled. On mono, InterpMethod serves as reference to the method to be called (which is embedded throughout the code) and later on, following compilation, it is filled with additional information relevant for execution, mainly the code pointer and size of the frame. On CoreCLR, this structure will likely lose some of its relevance, we will end up referencing MethodDesc's directly in interpreter instructions. In the future, it will probably make more sense to have a few common fields of the InterpMethod (like the frame size) stored as a code header to the interpreter IR, with a InterpMethod pointer to be stored as well in the code header for less frequent data.
        - set the CORINFO_FLG_INTERPRETER flag in the MethodDesc. This sets up the method to be called through the InterpreterStub.
        - some dummy data that is necessary by the runtime is being generated, mainly the code

The interpreter execution is included in the main coreclr library for convenience and speed. It will be contained in interpexec.h and interpexec.cpp while also including the interpretershared.h header from the interpreter library. This header will contain any defines and types that are relevant both during compilation as well as execution (for example the InterpMethod data or the opcode values). For the execution, InterpFrame is a structure on the native stack that is created for each interpreter frame. It will reference the method being executed as well as pointers to the interpreter stack. InterpThreadContext is a thread local structure that contains execution relevant information for the interpreter, for now it just maintains the position of the interpreter stack pointer.

BrzVlad committed:
Add support for compilation and execution of a simple method

This commit ports over some of the compilation machinery from mono. Method compilation takes place in GenerateCode. It allocates a basic block, goes over the IL code and emits code as it goes, while updating the IL stack state. Method code is kept as a linked list of basic blocks. Each basic block keeps a linked list of instructions. The instruction information is represented by an opcode, source and destination vars and additional instruction data. Once the IL code is imported, the var offset allocator needs to kick in in order to allocate an offset for each variable referenced in the code. Once offsets are allocated the final interpreter IR can be generated, writing the opcode and instruction data linearly in the code, while replacing variable indexes with the actual variable offset.

Unlike the mono interpreter, that maintains the code as an uint16_t stream, we chose to use from the start an int32_t based stream. The mono interpreter runs into limitations when executing very large methods, where some offsets don't fit in the uint16_t space. There will be a performance penalty for this that will be addressed in the distant future, by adding short versions for common opcodes.

jkotas commented on src/coreclr/interpreter/ee_il_dll.cpp:
```diff
@@ -0,0 +1,142 @@
+
+#include <vector>
+
+// FIXME We will probably end up not needing this table.
```
If this .dll should act like a special JIT, it should not have any global state. All global state should be on the VM side.

BrzVlad replied:
Wasn't really keen on further refactoring for this table which has a dubious use case anyway. I added another commit where I embed the `InterpMethod` pointer at the start of the code, making easier to obtain this information directly from compiled code, without further indirection through `MethodDesc`. In the future it might be best for it to be stored in some kind of header for interpreter code, probably once we have a different code manager for interpreter code.

kg commented on src/coreclr/inc/corjit.h:
```diff
@@ -104,6 +104,12 @@ class ICorJitInfo;
+class ICorInterpreter
+{
+    public:
+        virtual void* GetInterpMethod(CORINFO_METHOD_HANDLE methodHnd) = 0;
```
Do we want to make this an opaque strongly typed handle instead of void*?

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -0,0 +1,226 @@
+
+    int32_t m_totalVarsStackSize = 0;
+    int32_t m_paramAreaOffset = 0;
+    void    AllocVarOffsetCB(int *pVar, void *pData);
```
It's unclear from the name and signature of these methods what they do/how to use them, might be worth adding a comment later

kg commented on src/coreclr/interpreter/ee_il_dll.cpp:
```diff
@@ -0,0 +1,142 @@
+// If deemed useful, use some hashtable implementation instead.
+std::vector<std::pair<CORINFO_METHOD_HANDLE,InterpMethod*>> g_interpCodeHash;
+
+static InterpMethod* InterpGetInterpMethod(CORINFO_METHOD_HANDLE methodHnd)
```
Are there scenarios where we want to do a get without on-demand allocating a new one? Should this method have a 'create_if_missing' bool param?

kg commented on src/coreclr/interpreter/ee_il_dll.cpp:
```diff
@@ -0,0 +1,142 @@
+/*****************************************************************************/
+extern "C" DLLEXPORT void jitStartup(ICorJitHost* jitHost)
+{
+    if (g_interpInitialized)
```
should we check and throw if someone calls startup again with a different jit host?

kg commented on src/coreclr/interpreter/ee_il_dll.cpp:
```diff
@@ -0,0 +1,142 @@
+    InterpMethod *pMethod = InterpGetInterpMethod(methodInfo->ftn);
+    pMethod->compiled = true;
+ 
+    compHnd->setMethodAttribs(methodInfo->ftn, CORINFO_FLG_INTERPRETER);
```
Is this a race in the future if we set the INTERPRETER flag before we're done initializing all the interpreter state and metadata for the method? Or do we actually *want* to set it early so that once threading is wired up, other threads will see the flag and wait for us to finish compiling?

jkotas replied:
This flag should not be set by the interpreter JIT in the first place. The synchronization for multiple threads trying to JIT a method at the same and other global state management should be done on the VM side around the call to the interpreter JIT, like it is done for regular JIT.

BrzVlad replied:
Added a FIXME for now. Seems to me that this is something best refactored once we stop relying on the AltJit for kicking interpreter compilation.

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -0,0 +1,574 @@
+        ins->pNext->pPrev = ins;
+    }
+
+    ins->ilOffset = -1;
```
should NewIns be responsible for initializing this instead?

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -0,0 +1,574 @@
+    return InsertInsBB(m_pCBB, pPrevIns, opcode);
+}
+
+InterpInst* InterpCompiler::FirstIns(InterpBasicBlock *pBB)
```
I might want to rename this to FirstRealIns

kg commented:
One general note from reviewing is that we probably need handling for allocation failure in the interpreter. Since it has its own helpers for allocation it should be straightforward to centralize it and i.e. do a null check and throw/assert on allocation failure inside the helper so that the interp doesn't have to check for null every time it allocates.

On most targets attempting to deref a NULL allocation would fail fast but on wasm the interpreter would stomp all over memory, which isn't great, so we do need the explicit checks.

lateralusX commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -0,0 +1,226 @@
+{
+    int32_t opcode;
+    InterpInst *pNext, *pPrev;
+    int ilOffset;
```
should this be int32_t inline with types used on other members ?

lateralusX commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -0,0 +1,226 @@
+
+struct InterpBasicBlock
+{
+    int index;
```
maybe order it differently to reduce padding ?

lateralusX commented on src/coreclr/interpreter/ee_il_dll.cpp:
```diff
@@ -0,0 +1,146 @@
+#include <stdio.h>
+
+#ifdef _MSC_VER
+#define DLLEXPORT
```
Maybe its CoreClr convention to name things DLL on all platforms, but if that not a naming schema then maybe we could drop the dll naming and use INTERP_API or similar ? Maybe also drop the _dll from the file name if we would like to build this source as a static library in future.

jkotas replied:
This is copy&paste from https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/ee_il_dll.cpp . I agree that the historical dll names, etc. does not make sense. It would be nice to apply the same cleanup both in the regular JIT and the interpreter JIT so that both have the same shape.

lateralusX commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1838,6 +1838,18 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
+#ifdef FEATURE_INTERPRETER
+void InitializeInterpreter(IN HINSTANCE phJit, OUT ICorInterpreter **pNewInterpreter)
+{
+    typedef ICorInterpreter* (__stdcall* pGetInterpreterFn)();
```
if we mark this as __stdcall it should be synced with the actual implementation (that I believe didn't have any attributes).

jkotas replied:
Another copy&paste from the historic JIT/EE interface setup. It would be best to clean it up. For internal implementation details like this one, we prefer to omit explicit calling conventions.

BrzVlad committed:
Stop using a mapping table between MethodDesc and InterpMethod

InterpMethod* is used to handle interpreted method execution on mono, but it is not really needed. For CoreCLR we will turn this structure in a header to the compiled IR code, and identify other methods directly by the IR code pointer. The first pointer slot in the IR code will contain a pointer to the InterpMethod containing any other relevant information needed in the method execution.

janvorli commented on src/coreclr/interpreter/eeinterp.cpp:
```diff
@@ -0,0 +1,111 @@
+    int32_t *pIRCode = compiler.GetCode(&IRCodeSize);
+ 
+    // FIXME this shouldn't be here
+    compHnd->setMethodAttribs(methodInfo->ftn, CORINFO_FLG_INTERPRETER);
```
@jkotas these are special attributes that are used for communicating details back to the runtime. The JIT does the same thing. See e.g. the CORINFO_FLG_SWITCHED_TO_OPTIMIZED usage. These flags are special and are not set on the `MethodDesc`. I see reporting back that the JIT generated interpreted code using this flag as something in line with how the other JIT flags are used.

jkotas replied:
These special attributes are used for communicating details that that runtime cannot possibly know. They are computed from non-trivial analysis of the IL. This detail is not like that.

The runtime should know upfront that it is about to call JIT to generate interpreter code. This is related to our discussion about the dedicated code manager for interpreter code in the other PR.

Also, there is `getExpectedTargetArchitecture` JIT/EE interface API. This API should return a new unique architecture value that identifies interpreter. It requires runtime to know that we are generating interpreter code upfront as well.

jkotas replied:
> So the interpreter itself decides which methods it can compile and the runtime falls back to regular JIT for the ones that it rejects

`CORJIT_SKIPPED` return code from the JIT is meant for this purpose. The current altjit mechanism uses it to communicate that the aljit was not interested in compiling the method and regular JIT should be used as a fallback.

janvorli replied:
Yes, that's the way it works with the interpreter too. The reason I was mentioning this is that the runtime doesn't know if the AltJIT is another JIT or the interpreter. So the new attribute is a way to communicate that back to the caller that needs to know that without introducing something else for the runtime to recognize that the AltJIT produced interpreter IR byte code. We could come up with possibly many other ways to do that (setting a bit on the code address returned by the interpreter, adding a new API to the JIT to get such info etc.) However, the attribute seemed to be the least intrusive way to achieve that.

jkotas replied:
It is ok for now, but I am not convinced that it is the design we want in the long run.

janvorli commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -0,0 +1,583 @@
+{
+    int length = g_interpOpLen[opcode];
+    int numSVars = g_interpOpDVars[opcode];
+    int numDVars = g_interpOpDVars[opcode];
```
Is this correct? Both the numSVars and numDVars are set from the same array (g_interpOpDVars)

janvorli commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -0,0 +1,583 @@
+
+// FIXME Replace this
+static inline uint32_t ClzI4(uint32_t val)
+{
```
There is a `__builtin_clz` provided by Unix compilers and `_BitScanReverse` by MSVC that you can use. JIT uses them too (see BitOperations::BitScanReverse)

janvorli commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -0,0 +1,64 @@
+    if (!threadContext)
+    {
+        threadContext = new InterpThreadContext;
+        // FIXME valloc with INTERP_STACK_ALIGNMENT alignment
```
According to doc, `valloc` is an obsolete function and it aligns on page boundary. We should use `aligned_alloc` instead (once we decide to fix this "FIXME").

BrzVlad replied:
I wasn't even aware there was a function `valloc` in the C libraries. I meant that instead of using the malloc allocator we should allocate directly from OS with VirtualAlloc/mmap given this is a large chunk of memory. But I'm not yet familiar with the PAL stuff so I just added the comment.

janvorli replied:
Ah, got it.

janvorli commented on src/coreclr/interpreter/eeinterp.cpp:
```diff
@@ -95,33 +61,36 @@ CorJitResult CILInterp::compileMethod(ICorJitInfo*         compHnd,
     // FIXME this shouldn't be here
     compHnd->setMethodAttribs(methodInfo->ftn, CORINFO_FLG_INTERPRETER);
 
+    uint32_t sizeOfCode = sizeof(void*) + IRCodeSize * sizeof(int32_t);
```
A nit - sizeof(void*) -> sizeof(InterpMethod*) to make it clear?

janvorli replied:
I actually wonder - would there be any downside if  we stored the `InterpMethod` here directly instead of storing a pointer to it and allocating it separately?

BrzVlad replied:
As the implementation progresses the size of `InterpMethod` will likely become quite large, while for a lot of simple methods it will be unnecessary. I see some of the common contents of it being moved to the code header for faster access, while less common ones to be accessed through an optional `InterpMethod*` pointer in the header. For simplicity at the beginning I say we just keep this fixed pointer instead and adjust it later as the implementation progresses.

janvorli replied:
Ok, makes sense.

janvorli commented on src/coreclr/vm/interpexec.h:
```diff
@@ -0,0 +1,48 @@
+    } data;
+};
+
+struct InterpFrame
```
We will need to also have an explicit frame derived from the `Frame` class (in frames.h) created at the entry point to the interpreter when JITed / crossgened code calls interpreted method for the purpose of stack walking. I will add that one in my follow up change, but since those frames are all named "XXXFrame", it would be confusing to have the `InterpFrame` here named similarly. That was also one of the feedbacks that @kg has provided to one of the documents I've produced during the interpreter analysis experiment.
So I'd like this to be named somehow differently. My plan was to name the other one `InterpreterFrame`. What would you think about renaming this one to `InterpContext`?

janvorli reviewed (approved):
LGTM, thank you!

jkotas commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -0,0 +1,587 @@
+
+#include "openum.h"
+
+static StackType g_stackTypeFromInterpType[] =
```
const?

BrzVlad commented:
@jkotas Is this good to merge ?

jkotas commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -0,0 +1,226 @@
+    int32_t* m_pMethodCode;
+    int32_t m_MethodCodeSize; // in int32_t
+
+    void OptimizeCode();
```
```suggestion
```

jkotas commented on src/coreclr/interpreter/compileropt.cpp:
```diff
@@ -0,0 +1,45 @@
+    m_totalVarsStackSize = ALIGN_UP_TO(m_totalVarsStackSize, INTERP_STACK_ALIGNMENT);
+    m_paramAreaOffset = m_totalVarsStackSize;
+}
+
+void InterpCompiler::OptimizeCode()
+{
+    // TODO
+}
```
```suggestion
```
We should not be investing into code optimizations in this interpreter. Optimizing interpreter should be built from RyuJIT codebase and leverage all optimizations that RyuJIT does.

BrzVlad replied:
Optimization in general for the interpreter is something far away, at least 1 year, so no work will be done on it for the prototype anyway. I think reusing RyuJIT optimizations requires a complete redesign of the interpreter code gen so it has a similar way of representing all the compilation data. Doesn't seem very attractive to me, especially given we already have a set of decently advanced optimization on mono already that can plug in very easily. So it seems like much work for little gain at this moment.

jkotas replied:
> Doesn't seem very attractive to me, especially given we already have a set of decently advanced optimization on mono that can plug in very easily.

The primary goal of this project is to minimize our long-term costs. We do not want to be maintaining two advanced optimizers.

> I think reusing RyuJIT optimizations requires a complete redesign of the interpreter code gen so it has a similar way of representing all the compilation data.

Once we get there, we should look at this as building RyuJIT backend that produces interpreter byte code.

jkotas reviewed (commented):
> @jkotas Is this good to merge ?

Fine with me. I have not reviewed the changes in detail.

(left a minor comment)
